log-levels.md

```diff
@@ -0,0 +1,203 @@
+  - only log warnings about actionable things which the end-user of your library can do something about. Using the the "weird header detected" log statement as an example: it would not be a good candidate to log as a warning because the server developer has no way to fix the users of their service to stop sending weird headers, so the server should not be logging this information as a warning.
+
+Exceptions to the "avoid logging warnings" rule:
+
+- "background processes" such as tasks scheduled on a periodic timer, may not have any other means of communicating a failure or warning to the end user of the library other than through logging.
+  - consider offering an API that would collect errors at runtime, and then you can avoid logging errors manually. This can often take the form of a customizable "on error" hook that the library accepts when constructing the scheduled job. If the handler is not customized, we can log the errors, but if it was, it again is up to the end-user of the library to decide what to do with them.
+
+### Suggested logging style
+
+While libraries are free to use whichever logging message style they choose, here are some best practices to follow if you want users of your libraries *love* the logs your library produces.
+
+Firstly, it is important to remember that both the message of a log statement as well as the metadata in [swift-log](https://github.com/apple/swift-log) are [autoclosures](https://docs.swift.org/swift-book/LanguageGuide/Closures.html#ID543), which are only invoked if the logger has a log level set such that it must emit a message for the message given. As such, messages e.g. logged at `trace` do not "materialize" their string and metadata representation unless they are actually needed:
+
+```swift
+    public func debug(_ message: @autoclosure () -> Logger.Message,
+                      metadata: @autoclosure () -> Logger.Metadata? = nil,
+                      source: @autoclosure () -> String? = nil,
+                      file: String = #file, function: String = #function, line: UInt = #line) {
+```
+
+And a minor yet important hint: avoid inserting newlines and other control characters into log statements (!). Many log aggregation systems assume that a single line in a logged output is specifically "one log statement" which can accidentally break if we log not sanitized, potentially multi-line, strings. This isn't a problem for _all_ log backends, e.g. some will automatically sanitize and form a JSON payload with `{message: "..."}` before emitting it to a backend service collecting the logs, but plain old stream (or file) loggers usually assume that one line equals one log statement - it also makes grepping through logs more reliable.
+
+#### Logging with Correlation IDs / Trace IDs
```

Comment:
I think we should go into some detail regarding background/connection loggers for longer running services/clients. I mean the kind of that we inject into

- an [`HTTPClient` on startup](https://github.com/swift-server/async-http-client/blob/5d9b784d69e73c67ba3339933c5209943639d472/Sources/AsyncHTTPClient/HTTPClient.swift#L94)
- an [`AWSClient` on startup](https://github.com/soto-project/soto-core/blob/69d9bf8c4e9418ced238b4c41a2c805c7f3c4b2d/Sources/SotoCore/AWSClient.swift#L73)
- a [`PostgresConnection` on startup](https://github.com/vapor/postgres-nio/blob/d97f26858003d87a1d2b52c0061b243035620996/Sources/PostgresNIO/New/PSQLConnection.swift#L102)

It is worth mentioning that those long running instances in normal circumstances should not be started with a logger that already has a correlation id attached.

Reply:
> It is worth mentioning that those long running instances in normal circumstances should not be started with a logger that already has a correlation id attached.

That is specifically the reason swift-distributed-tracing logging context / baggage is designed like it is. Specifically, these are called "top level" logging contexts: https://apple.github.io/swift-distributed-tracing-baggage/docs/current/Baggage/Structs/DefaultLoggingContext.html#/s:7Baggage21DefaultLoggingContextV8topLevel6loggerAC0C06LoggerV_tFZ and they don't contain a baggage, as such, they would not contain correlation IDs.


Really, whenever we talk about any correlation IDs we should be thinking in terms of LoggingContext and tracing APIs. Using "just logger metadata" as carrying these is a workaround rather than the proper way IMHO.

Reply:
I can explain this a bit more here, thanks!